File: runtimes/android/host/src/main/kotlin/dev/flux/host/signal/SignalGraph.kt
package dev.flux.host.signal

import dev.flux.host.vm.FluxValue
import dev.flux.host.vm.SignalStore

/**
 * The reactive state of a single signal cell (ADR-0044, MLP v2 first-class async).
 *
 * - [Ready]: the cell holds a resolved value (`values[id]` is authoritative).
 * - [Pending]: an async-derived/resource cell has an in-flight future; `values[id]`
 *   may hold a stale `Ready` value or a placeholder — readers rendering the cell
 *   should surface the pending branch (e.g. the `when ... is_loading` form) rather
 *   than mutating the native view for the not-yet-resolved value.
 * - [Error]: the future resolved with a fault; [error] carries the message.
 */
public enum class CellState {
    Ready,
    Pending,
    Error,
}

/**
 * A fine-grained, dependency-tracked signal graph.
 *
 * Each signal holds a [FluxValue] in the VM's value representation (the VM is
 * the sole writer/reader; see [SignalStore]). Reads and writes are observed so
 * a handler dispatch can compute the minimal set of dependent signals to
 * propagate. The graph batches derived updates and replays them in topological
 * order, mirroring the Swift host's `SignalGraph` (FLUX-006).
 *
 * The MLP host is handler-authoritative: a tap runs a closure in
 * [dev.flux.host.vm.FluxBytecodeVM], which writes signals; the graph then
 * notifies subscribers for the changed ids (the eventual native view mutation).
 *
 * For MLP v2 async (ADR-0044), each cell also carries a [CellState] so async
 * derived/resource cells can be `Pending` while their future is in flight. The
 * graph does not mutate the native view for a `Pending` cell; it only flips the
 * state and notifies, letting the author's `when is_loading` branch drive the UI.
 */
public class SignalGraph : SignalStore {
    private val values = LinkedHashMap<UInt, FluxValue>()
    private val subscribers = LinkedHashMap<UInt, MutableSet<(FluxValue) -> Unit>>()
    private val pending = LinkedHashSet<UInt>()
    private val states = LinkedHashMap<UInt, CellState>()

    /** Returns the current value of [id], or `null` when unbound. */
    override fun read(id: UInt): FluxValue? = values[id]

    /** Returns the reactive [CellState] of [id], defaulting to [CellState.Ready]. */
    public fun cellState(id: UInt): CellState = states[id] ?: CellState.Ready

    /** Marks [id] as [CellState.Pending] (an async-derived/resource cell went in flight). */
    public fun markPending(id: UInt) {
        states[id] = CellState.Pending
    }

    /** Marks [id] as [CellState.Error] with [message]; the future resolved with a fault. */
    public fun markError(
        id: UInt,
        message: String,
    ) {
        states[id] = CellState.Error
    }

    /** Writes [value] into [id], recording it for the next [flush]. */
    override fun write(
        id: UInt,
        value: FluxValue,
    ) {
        values[id] = value
        // A successful write resolves any pending/error cell back to Ready.
        states[id] = CellState.Ready
        pending.add(id)
    }

    /** Returns every written signal as a sorted `(id, value)` list. */
    override fun snapshot(): List<Pair<UInt, FluxValue>> = values.entries.sortedBy { it.key }.map { (k, v) -> k to v }

    /** Seeds many signals at once (Init / StateDelta). */
    public fun seed(cells: Iterable<Pair<UInt, FluxValue>>) {
        for ((id, v) in cells) values[id] = v
    }

    /** Subscribes [block] to changes on [id]; returns a handle to unsubscribe. */
    public fun subscribe(
        id: UInt,
        block: (FluxValue) -> Unit,
    ): Subscription {
        val set = subscribers.getOrPut(id) { LinkedHashSet() }
        set.add(block)
        return Subscription { set.remove(block) }
    }

    /**
     * Propagates all pending writes to their subscribers, in ascending signal
     * id order (a deterministic topological order for the MLP graph). A [flush]
     * drains exactly the signals written since the last flush; concurrent
     * writes during propagation are appended to the next batch.
     */
    public fun flush() {
        if (pending.isEmpty()) return
        val batch = pending.sorted()
        pending.clear()
        for (id in batch) {
            val value = values[id] ?: continue
            subscribers[id]?.toList()?.forEach { it(value) }
        }
    }

    /** A subscription handle; [dispose] removes the callback. */
    public data class Subscription(
        val dispose: () -> Unit,
    )
}

File: runtimes/ios/FluxHost/Sources/FluxHost/SignalGraph.swift
//  SignalGraph.swift
//  SolidJS-style reactive signal graph (FLUX-006 scope item 8).
//
//  A `SignalGraph` owns a value-semantic map of `SignalId -> FluxValue` plus the
//  dependency edges between signals and derived computations.
//
//  Audit D7: writes commit values immediately (reads see latest state) but
//  notifications are batched and flushed in ascending signal-id order,
//  mirroring Kotlin's SignalGraph (pending + flush).

import Foundation

/// A signal identifier (matches the VM's `u32` signal space).
public typealias SignalId = UInt32

/// A token returned when subscribing; invalidating it removes the observer.
struct Subscription: Hashable, Sendable {
    let id: UInt64
}

/// The reactive state of a single signal cell (ADR-0044, MLP v2 first-class async).
public enum CellState: Equatable, Sendable {
    case ready
    case pending
    case error(message: String)
}

/// A minimal reactive signal store. Reads are O(1) dictionary lookups; writes
/// notify only the observers registered for the written signal. It conforms to
/// `SignalStore` so the VM can run handlers against it directly.
public struct SignalGraph: SignalStore {
    /// The current value of every signal.
    private(set) var values: [SignalId: FluxValue]
    /// Observers keyed by the signal they watch.
    private var observers: [SignalId: [Subscription: () -> Void]]
    /// Monotonic id source for subscriptions.
    private var nextSub: UInt64
    /// Monotonic id source for `allocateCell`, drawn from a high ceiling so it
    /// never collides with fixed ids like 97/99.
    private var nextCell: UInt32 = 1_000_000
    /// Reactive state of each cell (ADR-0044). A successful `write` resolves a
    /// cell back to `.ready`; async-derived/resource cells are `.pending` while
    /// their future is in flight.
    private var cellStates: [SignalId: CellState]
    /// Audit D7: pending notifications batched for ordered flush.
    private var pendingNotifications: Set<SignalId> = []

    /// Creates an empty graph.
    public init(values: [SignalId: FluxValue] = [:]) {
        self.values = values
        self.observers = [:]
        self.nextSub = 1
        self.cellStates = [:]
        self.nextCell = 1_000_000
    }

    /// Reads a signal's current value, or `nil` if it has never been written.
    public func read(_ id: UInt32) -> FluxValue? {
        values[id]
    }

    /// Returns the reactive [CellState] of `id`, defaulting to `.ready`.
    public func cellState(_ id: UInt32) -> CellState {
        cellStates[id] ?? .ready
    }

    /// Marks `id` as `.pending` (an async-derived/resource cell went in flight).
    public mutating func markPending(_ id: UInt32) {
        cellStates[id] = .pending
    }

    /// Marks `id` as `.error` with `message`; the future resolved with a fault.
    mutating func markError(_ id: UInt32, message: String) {
        cellStates[id] = .error(message: message)
    }

    /// Writes a value and records it for the next [flush].
    /// Audit D7: value is committed immediately (reads see latest state), but
    /// notifications are deferred until flush() to match Kotlin batching.
    public mutating func write(_ id: UInt32, _ value: FluxValue) {
        values[id] = value
        cellStates[id] = .ready
        pendingNotifications.insert(id)
        #if DEBUG
        fluxDevtoolsEmit(.signalWrite(signalId: id, oldValue: .null, newValue: value, triggeredEffectIds: []))
        #endif
    }

    /// Audit D7: flushes pending notifications in ascending signal-id order.
    /// Mirrors Kotlin's SignalGraph.flush().
    public mutating func flush() {
        if pendingNotifications.isEmpty { return }
        let batch = pendingNotifications.sorted()
        pendingNotifications.removeAll()
        for id in batch {
            let value = values[id] ?? .null
            let subs = observers[id] ?? [:]
            for notify in subs.values { notify() }
        }
    }

    /// Allocates a fresh, unbound signal id for a new capability result cell (ADR-0045).
    /// Drawn from a high ceiling so it never collides with fixed ids like 97/99.
    public mutating func allocateCell() -> UInt32 {
        nextCell &+= 1
        return nextCell
    }

    /// Resolves `id` to `value`, marking it `.ready` (an async capability finished).
    public mutating func resolveCell(_ id: UInt32, _ value: FluxValue) {
        values[id] = value
        cellStates[id] = .ready
    }

    /// Seeds a value without notifying observers (used for initial state seeds
    /// from an Init frame, where nothing is observing yet).
    mutating func seed(_ id: SignalId, _ value: FluxValue) {
        values[id] = value
    }

    /// Registers `observer` to run whenever `id` is written. Returns a
    /// `Subscription` that must be invalidated to stop observing.
    mutating func observe(_ id: SignalId, _ observer: @escaping () -> Void) -> Subscription {
        let sub = Subscription(id: nextSub)
        nextSub &+= 1
        observers[id, default: [:]][sub] = observer
        return sub
    }

    /// Removes an observer previously returned by `observe`.
    mutating func invalidate(_ sub: Subscription) {
        for key in observers.keys {
            observers[key]?.removeValue(forKey: sub)
        }
    }

    /// Every written signal as a sorted `(id, value)` list.
    public func snapshot() -> [(UInt32, FluxValue)] {
        values.map { ($0.key, $0.value) }.sorted { $0.0 < $1.0 }
    }
}
